### PR TITLE
WT-10727 Make the util dump command interactive

### PR DESCRIPTION
**Pull request by etienneptl**
When the -e option is given to the dump command, it is possible to go through the file in an interactive fashion:
```
$ ./wt dump -e file:test_dump.wt

**************************
Explore mode for file:test_dump.wt.
**************************
Enter 'h' for help, 'q' to exit.
```

### CODE REVIEW

File: src/utilities/util_dump.c [hunk 1]
```diff
@@ -144,10 +167,14 @@ util_dump(WT_SESSION *session, int argc, char *argv[])
     /* Open any optional output file. */
     if (ofile == NULL)
         fp = stdout;
-    else if ((fp = fopen(ofile, "w")) == NULL)
+    else if (explore) {
+        fprintf(stderr, "%s: the options are -e and -f are incompatible\n", progname);
```
Comment (lukech) [inline src/utilities/util_dump.c:171]: Typo for the 1st "are"?

File: src/utilities/util_dump.c [hunk 2]
```diff
@@ -726,6 +758,321 @@ dump_record(WT_CURSOR *cursor, const char *key, bool reverse, bool search_near,
     return (ret);
 }
 
+/*
+ * dump_explore --
+ *     Dump content in an interactive fashion.
```
Comment (lukech) [inline src/utilities/util_dump.c:763]: content -> contents (or data?)

File: src/utilities/util_dump.c (hunk 1 above)
Comment (etienneptl) [reply src/utilities/util_dump.c]: Fixed 1da526bb059bda937ae95bfaac45ede554db16d2

File: src/utilities/util_dump.c (hunk 2 above)
Comment (etienneptl) [reply src/utilities/util_dump.c]: Fixed f6edb3e2a7a4571e2a0c18a3f19ccbda76b13843

File: src/utilities/util_dump.c [hunk 3]
```diff
@@ -814,6 +847,321 @@ dump_all_records(WT_CURSOR *cursor, bool reverse, bool json)
...
+                ret = 0;
+                break;
+            }
+
+            if ((ret = cursor->get_value(cursor, &value)) != 0)
+                return (util_cerr(cursor, "get_value", ret));
+
+            printf("Current position:\n");
+            printf("key:%s\n", key);
+            printf("key:%s\n", value);
```
Comment (lukech) [inline src/utilities/util_dump.c:930]: key-> value

File: src/utilities/util_dump.c [hunk 4]
```diff
@@ -814,6 +847,321 @@ dump_all_records(WT_CURSOR *cursor, bool reverse, bool json)
...
+
+        first_arg = args[0];
+        switch (first_arg[0]) {
+        /* Cursor info. */
+        case 'a':
+            ret = cursor->get_key(cursor, &key);
+            if (ret != 0 && ret != EINVAL)
+                return (util_cerr(cursor, "get_key", ret));
+            else if (ret == EINVAL) {
+                printf("Error: the cursor needs to be positioned.\n");
```
Comment (lukech) [inline src/utilities/util_dump.c:920]: I'm curious in which case(s) one could run into an unpositioned cursor. Will running cursor->next to the end be one of the cases? Should we add the suggestion of running cursor reset "-c" in the error message?

File: src/utilities/util_dump.c [hunk 5]
```diff
@@ -814,6 +847,321 @@ dump_all_records(WT_CURSOR *cursor, bool reverse, bool json)
...
 }
 
+/*
+ * dump_explore --
+ *     Dump data in an interactive fashion.
+ */
+static int
+dump_explore(WT_CURSOR *cursor, const char *uri, bool reverse, bool pretty, bool hex, bool json)
+{
+#define MAX_LIMIT 256
```
Comment (marcbutler) [inline src/utilities/util_dump.c:857]: If feel this is really problematic. The name is vague, and there is no check to see if it the macro is already defined also it is not undef`ed at the end of the function.
It is used in multiple places to indicate the limits on multiple but logically independent arrays.

**Review by marcbutler — COMMENTED** [review]
First round to get you moving.

File: src/utilities/util_dump.c (hunk 4 above)
Comment (etienneptl) [reply src/utilities/util_dump.c:920]: This case happens when the first command is `a` or whenever the cursor is reset and the next command is `a`.

File: src/utilities/util_dump.c (hunk 3 above)
Comment (etienneptl) [reply src/utilities/util_dump.c]: Thanks! 0855211a4bd2e95445097ebc87560357e4a55dc2

File: src/utilities/util_dump.c [hunk 6]
```diff
@@ -814,6 +847,321 @@ dump_all_records(WT_CURSOR *cursor, bool reverse, bool json)
...
+                ret = (reverse ? cursor->prev(cursor) : cursor->next(cursor));
+            else
+                ret = (reverse ? cursor->next(cursor) : cursor->prev(cursor));
+            if (ret != 0 && ret != WT_NOTFOUND)
+                return (util_cerr(cursor, (reverse ? "prev" : "next"), ret));
+            if (ret == WT_NOTFOUND) {
+                printf("Error: %d\n", ret);
+                ret = 0;
+            } else {
+                if ((ret = cursor->get_key(cursor, &key)) != 0)
```
[the remark this answers is not in the document]
Comment (etienneptl) [reply src/utilities/util_dump.c]: Yes, done here: cb74c77e4614266081578fe07afe3f7f3a9fd0b0.

**Comment by luke-pearson** [comment]
I would have made this it's own option like `wt live` instead of a parameter to `wt dump`.

**Comment by luke-pearson** [comment]
So once skunkworks is complete will it be implemented fully?

File: src/utilities/util_dump.c [hunk 7]
```diff
@@ -814,6 +847,321 @@ dump_all_records(WT_CURSOR *cursor, bool reverse, bool json)
...
+                        break;
+                    }
+
+                    /* Set the cursor to the bookmark. */
+                    cursor->set_key(cursor, key);
+                    ret = cursor->search(cursor);
+                    if (ret != 0 && ret != WT_NOTFOUND)
+                        return (util_cerr(cursor, "search", ret));
+                    else if (ret == WT_NOTFOUND) {
+                        printf("Error: %d\n", ret);
```
[the remark this answers is not in the document]
Comment (etienneptl) [reply src/utilities/util_dump.c:963]: I would say, something obvious is that if a key is deleted and a bookmark was associated with it, we should delete the bookmark. I will address that.

File: src/utilities/util_dump.c (hunk 5 above)
Comment (etienneptl) [reply src/utilities/util_dump.c]: Not sure if f4e21cadec5526841818df0750baf25af80289b6 is better, let me know.

File: src/utilities/util_dump.c [hunk 8]
```diff
@@ -805,15 +842,328 @@ dump_all_records(WT_CURSOR *cursor, bool reverse, bool json)
...
+ */
+static int
+dump_explore(WT_CURSOR *cursor, const char *uri, bool reverse, bool pretty, bool hex, bool json)
+{
+    WT_DECL_RET;
+    WT_SESSION *session;
+    uint64_t bookmark_index, window;
+    int i, exact, num_args;
+    char *args[MAX_ARGS], *bookmarks[MAX_BOOKMARKS];
+    char *first_arg, user_input[MAX_ARGS], *current_arg;
```
Comment (marcbutler) [inline src/utilities/util_dump.c:869]: Possibly based on the naming should this be something like?:
```suggestion
    char *first_arg, user_input[ARG_BUF_SIZE], *current_arg;
```

File: src/utilities/util_dump.c [hunk 9]
```diff
@@ -805,15 +842,328 @@ dump_all_records(WT_CURSOR *cursor, bool reverse, bool json)
...
+    const char *key, *value;
+    bool once, search_near;
+
+    session = cursor->session;
+    once = search_near = false;
+    i = exact = num_args = 0;
+    bookmark_index = window = 0;
+    for (i = 0; i < MAX_BOOKMARKS; ++i)
```
Comment (marcbutler) [inline src/utilities/util_dump.c:877]:
```suggestion
    memset(bookmarks, 0, sizeof(bookmarks));
```

File: src/utilities/util_dump.c [hunk 10]
```diff
@@ -805,15 +842,328 @@ dump_all_records(WT_CURSOR *cursor, bool reverse, bool json)
...
+        once = true;
+
+        /* Time to exit. */
+        if (strlen(user_input) == 1 && user_input[0] == 'q')
+            break;
+
+        /* Parse the input. */
+        current_arg = strtok(user_input, " ");
+        while (current_arg != NULL) {
+            args[i] = malloc(ARG_BUF_SIZE);
```
Comment (marcbutler) [inline src/utilities/util_dump.c:902]: This leaks memory (no `free`).
I'd recommend args[i] actually be an array of strings of fixed size.
If an argument is too long for the fixed size buffer it is a parsing error.

File: src/utilities/util_dump.c [hunk 11]
```diff
@@ -805,15 +842,328 @@ dump_all_records(WT_CURSOR *cursor, bool reverse, bool json)
...
+            if ((ret = cursor->get_value(cursor, &value)) != 0)
+                return (util_cerr(cursor, "get_value", ret));
+
+            printf("Current position:\n");
+            printf("key:%s\n", key);
+            printf("value:%s\n", value);
+            break;
+        /* Bookmarks. */
+        case 'b':
+            if (strcmp(first_arg, "b") == 0) {
```
Comment (marcbutler) [inline src/utilities/util_dump.c:930]: I don't understand the need for this `strcmp` as `b` is matched in the case stmt.

File: src/utilities/util_dump.c [hunk 12]
```diff
@@ -805,15 +842,328 @@ dump_all_records(WT_CURSOR *cursor, bool reverse, bool json)
...
+                if (ret != 0 && ret != EINVAL)
+                    return (util_cerr(cursor, "get_key", ret));
+                else if (ret == EINVAL) {
+                    printf("Error: the cursor needs to be positioned to create a bookmark.\n");
+                    ret = 0;
+                    break;
+                }
+                for (i = 0; i < MAX_BOOKMARKS; ++i) {
+                    if (bookmarks[i] == NULL) {
+                        if ((bookmarks[i] = malloc(strlen(key))) == NULL)
```
Comment (marcbutler) [inline src/utilities/util_dump.c:999]: This leaks memory when the function exits. While I realize the util itself is probably going to exit. Still I suggest cleaning up the memory or providing a comment as to why it doesn't matter.

File: src/utilities/util_dump.c [hunk 13]
```diff
@@ -805,15 +842,328 @@ dump_all_records(WT_CURSOR *cursor, bool reverse, bool json)
...
+                }
+                /* Jump to the bookmark. */
+                else {
+                    if (util_str2num(session, args[1], true, &bookmark_index) != 0 ||
+                      bookmark_index >= MAX_BOOKMARKS) {
+                        printf("Error: please indicate a value between 0 and %d\n", MAX_BOOKMARKS);
+                        break;
+                    }
+                    key = bookmarks[bookmark_index];
+                    if (key == NULL) {
```
Comment (marcbutler) [inline src/utilities/util_dump.c:947]: There is nothing wrong here, but in the code I've read in the code base to date the following appears to be preferred:
```suggestion
                    if ((key = bookmarks[bookmark_index]) == NULL) {
```

File: src/utilities/util_dump.c [hunk 14]
```diff
@@ -805,15 +842,328 @@ dump_all_records(WT_CURSOR *cursor, bool reverse, bool json)
...
         return (util_err(session, EIO, NULL));
     return (0);
 }
 
+/*
+ * dump_explore --
+ *     Dump data in an interactive fashion.
+ */
+static int
+dump_explore(WT_CURSOR *cursor, const char *uri, bool reverse, bool pretty, bool hex, bool json)
```
Comment (marcbutler) [inline src/utilities/util_dump.c:862]: The extended switch statement in this function. Perhaps some functionality like common bookmark operations could be factored out?

File: src/utilities/util_dump.c (hunk 10 above)
Comment (etienneptl) [reply src/utilities/util_dump.c:902]: I agree with the `free`.
However, concerning the size, it is hard to have a fixed one especially if the user gives a key or a value.

File: src/utilities/util_dump.c (hunk 11 above)
Comment (etienneptl) [reply src/utilities/util_dump.c:930]: I am checking if this is exactly `b` as we could have, for instance, `bd`.

File: src/utilities/util_dump.c (hunk 9 above)
Comment (etienneptl) [reply src/utilities/util_dump.c]: Thanks, see 3a8666c193a870d2a071dd349105c98635176418.

File: src/utilities/util_dump.c (hunk 10 above)
Comment (etienneptl) [reply src/utilities/util_dump.c:902]: Updated the `free` in 3a8666c193a870d2a071dd349105c98635176418.

File: src/utilities/util_dump.c (hunk 12 above)
Comment (etienneptl) [reply src/utilities/util_dump.c]: Updated in 3a8666c193a870d2a071dd349105c98635176418, on failure we will jump to the `err` label that will do the cleanup.

File: src/utilities/util_dump.c (hunk 13 above)
Comment (etienneptl) [reply src/utilities/util_dump.c]: Updated in 3a8666c193a870d2a071dd349105c98635176418

**Review by lukech — APPROVED** [review]
LGTM. Thanks

File: src/utilities/util_dump.c (hunk 14 above)
Comment (etienneptl) [reply src/utilities/util_dump.c:968]: Yes, done in 5362a30267ea3bd6146bf2bf2370895c1b687d6a. Let me know what you think.

**Review by marcbutler — APPROVED** [review]
lgtm

**Pull request merged**